examples/tutorial/gl/fireworks.py doesn't work on OS X

Shows a black screen on any backend...

Still get:

```
    def glGetProgramInfoLog(program):
        res = GL.glGetProgramInfoLog(program)
>       return res.decode('utf-8')
E       AttributeError: 'str' object has no attribute 'decode'
```

as well as the ton of `max_corr` errors. The other two are gone though!

--- a/codegen/annotations.py
+++ b/codegen/annotations.py
@@ -500,7 +500,7 @@ def getProgramInfoLog(program):
     return infolog[:length[0]].decode('utf-8')
     # --- pyopengl
     res = GL.glGetProgramInfoLog(program)
-    return res.decode('utf-8')
+    return res.decode('utf-8') if isinstance(res, bytes) else res
 
 def getShaderInfoLog(shader):
     # --- gl es
@@ -511,7 +511,7 @@ def getShaderInfoLog(shader):
     return infolog[:length[0]].decode('utf-8')
     # --- pyopengl
     res = GL.glGetShaderInfoLog(shader)
-    return res.decode('utf-8')
+    return res.decode('utf-8') if isinstance(res, bytes) else res
 
 def getProgramParameter(program, pname):
     # --- gl es

--- a/examples/tutorial/gl/cube.py
+++ b/examples/tutorial/gl/cube.py
@@ -124,7 +124,6 @@ class Canvas(app.Canvas):
         app.Canvas.__init__(self, size=(512, 512),
                             title='Rotating cube (GL version)',
                             keys='interactive')
-        self.timer = app.Timer('auto', self.on_timer)
 
     def on_initialize(self, event):
         # Build & activate cube program
@@ -201,8 +200,8 @@ class Canvas(app.Canvas):
         # OpenGL initalization
         gl.glClearColor(0.30, 0.30, 0.35, 1.00)
         gl.glEnable(gl.GL_DEPTH_TEST)
-
-        self.timer.start()
+        self._resize(*(self.size + self.physical_size))
+        self.timer = app.Timer('auto', self.on_timer, start=True)
 
     def on_draw(self, event):
         gl.glClear(gl.GL_COLOR_BUFFER_BIT | gl.GL_DEPTH_BUFFER_BIT)
@@ -210,8 +209,10 @@ class Canvas(app.Canvas):
                           gl.GL_UNSIGNED_INT, None)
 
     def on_resize(self, event):
-        width, height = event.size
-        gl.glViewport(0, 0, width, height)
+        self._resize(*(event.size + event.physical_size))
+
+    def _resize(self, width, height, physical_width, physical_height):
+        gl.glViewport(0, 0, physical_width, physical_height)
         projection = perspective(35.0, width / float(height), 2.0, 10.0)
         loc = gl.glGetUniformLocation(self.cube, "u_projection")
         gl.glUniformMatrix4fv(loc, 1, False, projection)

--- a/examples/tutorial/gl/fireworks.py
+++ b/examples/tutorial/gl/fireworks.py
@@ -24,6 +24,8 @@ from vispy.gloo import gl
 
 
 vertex_code = """
+#version 120
+
 uniform float time;
 uniform vec3 center;
 attribute float lifetime;
@@ -44,6 +46,8 @@ void main () {
 """
 
 fragment_code = """
+#version 120
+
 uniform vec4 color;
 varying float v_lifetime;
 void main()

--- a/examples/tutorial/gl/fireworks.py
+++ b/examples/tutorial/gl/fireworks.py
@@ -60,7 +60,6 @@ class Canvas(app.Canvas):
     def __init__(self):
         app.Canvas.__init__(self, size=(800, 600), title='GL Fireworks',
                             keys='interactive')
-        self.timer = app.Timer('auto', self.on_timer)
 
     def on_initialize(self, event):
         # Build & activate program
@@ -113,8 +112,9 @@ class Canvas(app.Canvas):
         gl.glBlendFunc(gl.GL_SRC_ALPHA, gl.GL_ONE)
         gl.glEnable(34370)  # gl.GL_VERTEX_PROGRAM_POINT_SIZE
         gl.glEnable(34913)  # gl.GL_POINT_SPRITE
+        gl.glViewport(0, 0, *self.physical_size)
         self.new_explosion()
-        self.timer.start()
+        self.timer = app.Timer('auto', self.on_timer, start=True)
 
     def on_draw(self, event):
         gl.glClear(gl.GL_COLOR_BUFFER_BIT | gl.GL_DEPTH_BUFFER_BIT)

--- a/examples/tutorial/gl/quad.py
+++ b/examples/tutorial/gl/quad.py
@@ -108,7 +108,7 @@ class Canvas(app.Canvas):
         gl.glDrawArrays(gl.GL_TRIANGLE_STRIP, 0, 4)
 
     def on_resize(self, event):
-        gl.glViewport(0, 0, *event.size)
+        gl.glViewport(0, 0, *event.physical_size)
 
 if __name__ == '__main__':
     c = Canvas()

--- a/vispy/app/backends/_glfw.py
+++ b/vispy/app/backends/_glfw.py
@@ -364,7 +364,6 @@ class CanvasBackend(BaseCanvasBackend):
             # glfw.glfwSetWindowShouldClose()  # Does not really cause a close
             self._vispy_set_visible(False)
             self._id, id_ = None, self._id
-            glfw.glfwPollEvents()
             glfw.glfwDestroyWindow(id_)
             gc.collect()  # help ensure context gets destroyed
 

--- a/vispy/app/backends/_glfw.py
+++ b/vispy/app/backends/_glfw.py
@@ -393,8 +393,8 @@ class CanvasBackend(BaseCanvasBackend):
     def _on_resize(self, _id, w, h):
         if self._vispy_canvas is None:
             return
-        self._vispy_canvas.events.resize(size=(w, h), physical_size=
-                                         self._vispy_get_physical_size())
+        self._vispy_canvas.events.resize(
+            size=(w, h), physical_size=self._vispy_get_physical_size())
 
     def _on_close(self, _id):
         if self._vispy_canvas is None:

--- a/vispy/app/canvas.py
+++ b/vispy/app/canvas.py
@@ -108,7 +108,7 @@ class Canvas(object):
         if dpi is None:
             dpi = util_config['dpi']
         if dpi is None:
-            dpi = get_dpi()
+            dpi = get_dpi(raise_error=False)
         self.dpi = dpi
 
         # Create events

--- a/vispy/app/tests/test_simultaneous.py
+++ b/vispy/app/tests/test_simultaneous.py
@@ -80,7 +80,7 @@ def test_multiple_canvases():
             t = Timer(0.1, app=app, connect=on_timer, iterations=1,  # noqa
                       start=True)
             app.process_events()
-            sleep(0.2)
+            sleep(0.5)  # long for slow systems
             app.process_events()
             assert timer_ran
 

--- a/vispy/app/tests/test_simultaneous.py
+++ b/vispy/app/tests/test_simultaneous.py
@@ -67,8 +67,8 @@ def test_multiple_canvases():
             while (ct[0] < n_check or ct[1] < n_check) and time() < timeout:
                 app.process_events()
             print((ct, n_check))
-            assert n_check <= ct[0] <= n_check + 1
-            assert n_check <= ct[1] <= n_check + 1
+            assert n_check <= ct[0] <= n_check + 2  # be a bit lenient
+            assert n_check <= ct[1] <= n_check + 2
 
             # check timer
             global timer_ran
@@ -81,6 +81,7 @@ def test_multiple_canvases():
                       start=True)
             app.process_events()
             sleep(0.5)  # long for slow systems
+            app.process_events()
             app.process_events()
             assert timer_ran
 

--- a/vispy/app/timer.py
+++ b/vispy/app/timer.py
@@ -104,7 +104,14 @@ class Timer(object):
         emitting that number of events. If unspecified, then
         the previous value of self.iterations will be used. If the value is
         negative, then the timer will continue running until stop() is called.
+
+        If the timer is already running when this function is called, nothing
+        happens (timer continues running as it did previously, without
+        changing the interval, number of iterations, or emitting a timer
+        start event).
         """
+        if self.running:
+            return  # don't do anything if already running
         self.iter_count = 0
         if interval is not None:
             self.interval = interval

--- a/vispy/gloo/gl/_pyopengl2.py
+++ b/vispy/gloo/gl/_pyopengl2.py
@@ -129,7 +129,7 @@ def glGetRenderbufferParameter(target, pname):
 
 def glGetShaderInfoLog(shader):
     res = GL.glGetShaderInfoLog(shader)
-    return res.decode('utf-8')
+    return res.decode('utf-8') if isinstance(res, bytes) else res
 
 
 def glGetShaderSource(shader):

--- a/vispy/gloo/gl/_pyopengl2.py
+++ b/vispy/gloo/gl/_pyopengl2.py
@@ -117,7 +117,7 @@ def glGetFramebufferAttachmentParameter(target, attachment, pname):
 
 def glGetProgramInfoLog(program):
     res = GL.glGetProgramInfoLog(program)
-    return res.decode('utf-8')
+    return res.decode('utf-8') if isinstance(res, bytes) else res
 
 
 def glGetRenderbufferParameter(target, pname):

--- a/vispy/util/dpi/_linux.py
+++ b/vispy/util/dpi/_linux.py
@@ -8,6 +8,7 @@
 import re
 from subprocess import CalledProcessError
 
+from ..logs import logger
 from ..wrappers import run_subprocess
 
 
@@ -25,8 +26,18 @@ def _get_dpi_from(cmd, pattern, func):
             return func(*map(float, match.groups()))
 
 
-def get_dpi():
-    """Get screen DPI from the OS.
+def get_dpi(raise_error=True):
+    """Get screen DPI from the OS
+
+    Parameters
+    ----------
+    raise_error : bool
+        If True, raise an error if DPI could not be determined.
+
+    Returns
+    -------
+    dpi : float
+        Dots per inch of the primary screen.
     """
 
     from_xdpyinfo = _get_dpi_from(
@@ -40,5 +51,8 @@ def get_dpi():
         lambda x_px, y_px, x_mm, y_mm: 25.4 * (x_px / x_mm + y_px / y_mm) / 2)
     if from_xrandr is not None:
         return from_xrandr
-
-    raise RuntimeError('could not determine DPI')
+    if raise_error:
+        raise RuntimeError('could not determine DPI')
+    else:
+        logger.warning('could not determine DPI')
+    return 96

--- a/vispy/util/dpi/_quartz.py
+++ b/vispy/util/dpi/_quartz.py
@@ -7,8 +7,19 @@
 from ...ext.cocoapy import quartz
 
 
-def get_dpi():
-    """Get screen DPI from the OS"""
+def get_dpi(raise_error=True):
+    """Get screen DPI from the OS
+
+    Parameters
+    ----------
+    raise_error : bool
+        If True, raise an error if DPI could not be determined.
+
+    Returns
+    -------
+    dpi : float
+        Dots per inch of the primary screen.
+    """
     display = quartz.CGMainDisplayID()
     mm = quartz.CGDisplayScreenSize(display)
     px = quartz.CGDisplayBounds(display).size

--- a/vispy/util/dpi/_win32.py
+++ b/vispy/util/dpi/_win32.py
@@ -8,8 +8,19 @@ from ...ext.gdi32plus import (gdi32, user32, HORZSIZE, VERTSIZE,
                               HORZRES, VERTRES)
 
 
-def get_dpi():
-    """Get screen DPI from the OS"""
+def get_dpi(raise_error=True):
+    """Get screen DPI from the OS
+
+    Parameters
+    ----------
+    raise_error : bool
+        If True, raise an error if DPI could not be determined.
+
+    Returns
+    -------
+    dpi : float
+        Dots per inch of the primary screen.
+    """
     try:
         user32.SetProcessDPIAware()
     except AttributeError:

--- a/vispy/util/fetching.py
+++ b/vispy/util/fetching.py
@@ -274,7 +274,7 @@ def _fetch_file(url, file_name, print_destination=True):
     n_try = 3
     for ii in range(n_try):
         try:
-            data = urllib.request.urlopen(url, timeout=5.)
+            data = urllib.request.urlopen(url, timeout=15.)
         except Exception as e:
             if ii == n_try - 1:
                 raise RuntimeError('Error while fetching file %s.\n'

